pageserver/src/layered_repository.rs

```diff
@@ -1105,71 +1134,90 @@ impl LayeredTimeline {
-        // It would be smarter to only flush out in-memory layers that
-        // have accumulated a fair amount of changes. Note that the
-        // start and end LSNs of snapshot files belonging to different
-        // relations don't have to line up, although currently they do
-        // because of the way this works. So you could have a snapshot
-        // file covering LSN range 100-200 for one relation, and a
-        // snapshot file covering 150-250 for another relation. The
-        // read functions should even cope with snapshot files
-        // covering overlapping ranges for the same relation, although
-        // that situation never arises currently.
-        //
-        // Note: We aggressively freeze and unload all the layer
-        // structs. Even if a layer is actively being used. This
-        // keeps memory usage in check, but is probably too
-        // aggressive. Some kind of LRU policy would be appropriate.
+        // Take the in-memory layer with the oldest WAL record. If it's older
+        // than the threshold, write it out to disk as a new snapshot file.
+        // Repeat until all remaining in-memory layers are within the threshold.
         //
+        // That's necessary to limit the amount of WAL that needs to be kept
+        // in the safekeepers, and that needs to be reprocessed on page server
+        // crash. TODO: It's not a great policy for keeping memory usage in
+        // check, though. We should also aim at flushing layers that consume
+        // a lot of memory and/or aren't receiving much updates anymore.
+        let mut disk_consistent_lsn = last_record_lsn;
+        while let Some(oldest_layer) = layers.get_oldest_open_layer() {
+            let oldest_pending_lsn = oldest_layer.get_oldest_pending_lsn();
+            let distance = last_valid_lsn.0 - oldest_pending_lsn.0;
+            if !force && distance < OLDEST_INMEM_DISTANCE {
+                info!(
+                    "the oldest layer is now {} which is {} bytes behind last_valid_lsn",
+                    oldest_layer.get_seg_tag(),
+                    distance
+                );
+                disk_consistent_lsn = oldest_pending_lsn;
+                break;
+            }
 
-        // Call freeze() on any unfrozen layers (that is, layers that haven't
-        // been written to disk yet).
-        // Call unload() on all frozen layers, to release memory.
-        let mut iter = layers.iter_open_layers();
-        while let Some(layer) = iter.next() {
-            let (new_historic, new_open) = layer.freeze(last_valid_lsn, &self)?;
+            let (new_historic, new_open) = oldest_layer.freeze(last_valid_lsn, &self)?;
 
             // replace this layer with the new layers that 'freeze' returned
-            // (removes it if new_open is None)
-            iter.replace(new_open);
-
+            layers.pop_oldest();
+            if let Some(n) = new_open {
```

Comment:
Can we use something like `maybe_new_open` and `new_open`

Reply:
I see, it's confusing that `new_historic` is an Option while `new_open` is not.

I played with different alternatives, and wasn't very happy with any. 'maybe_new_open' sounds weird to me. One idea was to add explicit type markers to the 'let' clause, something like:

```
let (new_historic: Arc<SnapshotLayer>,
     new_open: Option<Arc<InMemoryLayer>>) = oldest_layer.freeze(last_valid_lsn, &self)?;
```
but Rust doesn't actually allow that, and that's not very pretty either.
 
I moved the 'trace' call inside 'freeze' function, that helps somewhat by just making this piece of code shorter.